spur/ssh.py

```diff
@@ -80,7 +80,14 @@ def generate_run_command(self, command_args, store_pid,
             commands.append("echo $$")
 
         if cwd is not None:
-            commands.append("cd {0}".format(escape_sh(cwd)))
+            # 1. Change directory before to check if the command exists
+            # (using command || which) because it might be in this
+            # directory:
+            # $ shell.run(["./foo"], cwd="/some/dir")
+            # 2. If we exit when this cd fails, we can't capture the which_return_code.
+            # To minimize the changes, we run cd again just before exec.
+            # Send stderr to /dev/null to capture only the second cd.
+            commands.append("cd {0} 2>/dev/null".format(escape_sh(cwd)))
```

Comment:
If we just run:
```
$ cd '/some/path' 2>/dev/null; echo$?; command -v 'ls' > /dev/null 2>&1 ||
which 'ls' > /dev/null 2>&1; echo $?; exec 'ls'
```
sure, we can check if changing directory succeeded in our Python code and raise an exception, but on the remote server, the full command will be executed anyway. That's what we want to prevent.
To do that we should use `&&`between commands to be sure next one is executed only if the previous one succeeded.